core/Network/TLS/Handshake/Server.hs

```diff
@@ -795,17 +784,19 @@ doHandshake13 sparams cred@(certChain, _) ctx chosenVersion usedCipher exts used
                 sendNewSessionTicket masterSecret rtt
               else
                 decryptError "cannot verify finished"
-        finishedAction hs = unexpected (show hs) (Just "finished 13")
+        expectFinished hs = unexpected (show hs) (Just "finished 13")
 
-        endOfEarlyDataAction hs@EndOfEarlyData13 = do
+    let expectEndOfEarlyData hs@EndOfEarlyData13 = do
             processHandshake13 ctx hs
             setRxState ctx usedHash usedCipher clientHandshakeTrafficSecret
-        endOfEarlyDataAction hs = unexpected (show hs) (Just "end of early data")
+        expectEndOfEarlyData hs = unexpected (show hs) (Just "end of early data")
 
-    if rtt0OK then do
-        setPendingActions ctx [endOfEarlyDataAction, finishedAction]
+    if serverWantClientCert sparams then
+        setPendingActions ctx [expectCertificate, expectCertVerify, expectFinished]
+      else if rtt0OK then
+        setPendingActions ctx [expectEndOfEarlyData, expectFinished]
       else do
-        setPendingActions ctx [finishedAction]
+        setPendingActions ctx [expectFinished]
```

Comment:
> `serverWantClientCert` should be ignored when resuming a session (PSK or 0-RTT). In these modes, the server does not send the certificate request, so should not expect to receive Certificate and CertVerify.

Done.

Reply:
> Additionally, if the client has no certificate to send, it will send Certificate message with an empty content and no CertVerify message. Implementation should handle this scenario as well.

OK. I checked RFC8446 again and understood the structure of handshake messages in this case. However, I don't know how servers should behave.

Reply:
We can do similarly to TLS12 with processClientCertificate and processCertificateVerify.

However it's not possible to use three independent pending actions, two needs to be merged. Another possibility is not to use pending actions at all for this part. They are only needed in case of 0-RTT handshake.

Reply:
Although not tested, I believe that 31ee5e579a9d5ae83ad55803d2f9eed8f66bedb8 implements certverify skipping. What I don't understand is which code rejects null cert-chain of client if the server dislikes it. Does `clientCertificate` implement such local policy?

Reply:
Deciding about empty certificate chain is the job of `onClientCertificate`.

Popping the action will work but it's not a very clean solution. Can't we just not use pending actions instead? From a user perspective it may be confusing to have invocation of callbacks and possible authentication failures transferred from `handshake` to `recvData` based on the client version. Since it is not post-handshake authentication.